### ISSUE TITLE
LNK1104 error when after a few minutes. LIB & INCLUDE vars and compiling not working

### ISSUE THREAD

**Issue by AnthonyTheKoala** [davisking/dlib#488]
My goal is to be able to go into python's IDLE and say 
<pre> import dlib as boo 
Traceback (most recent call last):
  File "<pyshell#0>", line 1, in <module>
    import dlib as boo
ImportError: No module named 'dlib'
</pre>. 
But before I can do this, I have been having frustrating moments for many hours. These are the steps taken to replicate the problem.
(1) unzip boost_1_63_0 to to c:\
<pre>c:\boost_1_63_0</pre>
(2) unzip dlib-19.4 to c:\
<pre>c:\dlib-19.4</pre>
(3) Install MSVC 15 cpp tools, visualcppbuildtools_full.exe, its update 3 vs2015.3.exe and update to update 3 vs14-kb3165756.exe to satisfy the requirement of v19.24215.1 for CL.exe.
(4) In Win7 64, set the following environment paths by Start (button) -> Computer -> System Properties -> Advanced System Settings -> Environment Variables -> 'creating' new variables in both New for A (user A) and New for System Variables.
In both cases, LIB and INCLUDE and path are in User A and System Variables.

Remark, reference to 8.1 in subfolders. It is backwards compatible with Win 7, https://developer.microsoft.com/en-us/windows/downloads/windows-8-1-sdk 

LIB
<pre>
"C:\Program Files (x86)\Windows Kits\8.1\Lib\winv6.3\um\x64";"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\Lib\um\x64";"C:\Program Files (x86)\Windows Kits\8.1\Lib\winv6.3\um\x64";"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\Lib\um\x64";"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\atlmfc\lib\amd64";"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib\amd64";"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib\store\amd64";"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib\store\references";
</pre>

INCLUDE
<pre>
"C:\Program Files (x86)\Windows Kits\8.1\Include\um";"C:\Program Files (x86)\Windows Kits\8.1\Include\shared";"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\Include\um";"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\atlmfc\include";"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include";
</pre>

path - a snippet of it - does not include other MSVC installations.
<pre>
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64;C:\Program Files (x86)\Windows Kits\8.1\bin\x64;%LIB%;%INCLUDE%;%BOOST_ROOT%;%BOOST_LIBRARYDIR%;%BOOST_LIBRARYDIR%;%BOOST_ROOT%\bin.v2;</pre>

Other paths %BOOST_ROOT% and %BOOST_LIBRARYDIR% are variables
<pre>
BOOST_ROOT 
c:\boost_1_63_0

BOOST_LIBRARYDIR
c:\boost_1_63_0\stage\lib

BOOST_INCLUDE
c:\boost_1_63_0\boost    
</pre>

(4) Boost - bootstrap, b2 and bjam
Run a cmd console with admin rights and do vcvarsall.bat amd64, or open a MSVC 2015 C++ native build compiler for x64 with admin rights. 
<pre>
cd c:\boost_1_63_0
bootstrap
b2 --toolset=msvc address-model=64
bjam --with-python --toolset=msvc-2015 address-model=64 variant=release
...patience...
...patience...
...patience...
...found 3271 targets...

The Boost C++ Libraries were successfully built
</pre>

(5) DLIB - building and installing. I have successfully installed the examples, this is not relevant. I show what I did to install dlib. Again in a cmd console with vcvarsall.bat amd64 or opening an MSVC 2015 C++ native build compiler console
<pre>
cd c:\dlib-19.4
python setup.py install -G "Visual Studio 14 2015 Win64"
</pre>

Results:
Here are snippets from results.txt produced by python setup.py -G "Visual Studio 14 2015 Win64" > results.txt
First, the compiler said that boost was not found, YET, it managed to find the boost libraries
<pre>
-- Could NOT find Boost
CMake Warning at C:/Program Files/CMake/share/cmake-3.8/Modules/FindBoost.cmake:1520 (message):
  No header defined for python3; skipping header check
Call Stack (most recent call first):
  C:/dlib-19.4/dlib/cmake_utils/add_python_module:66 (FIND_PACKAGE)
  CMakeLists.txt:6 (include)
-- Boost version: 1.63.0
-- Found the following Boost libraries:
--   python3
-- Found PythonLibs: C:/Python34/libs/python34.lib (found suitable version "3.4.4", minimum required is "3.4")
-- USING BOOST_LIBS: optimized;C:/boost_1_63_0/stage/lib/libboost_python3-vc140-mt-s-1_63.lib;debug;C:/boost_1_63_0/stage/lib/libboost_python3-vc140-mt-sgd-1_63.lib
-- USING PYTHON_LIBS: C:/Python34/libs/python34.lib
</pre>

Second- there were a number of C4244 warnings relating to converting 'Py_ssize_t' to 'unsigned int'. These were for 'creating' *.cpp files for matrix, svm_ctrainer, svm_rank, shape predictor, just to name a few. There were 18 instances of C4244 warnings.
Typically, 
<pre>
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  matrix.cpp
</pre>

Second - there was a LNK1104 error. The file 'libboost_python-vc140-mt-s-1_63.lib' could not be opened. This file is in c:\boost_1_63_0\stage\lib. I don't know why. Looking at the c:\boost_1_63_0\stage
After the line dlib_.dir\Release\gui.obj the compiler was 'stuck' for a few minutes until the LNK1104 error appeared. 
<pre>
  dlib_.dir\Release\gui.obj     
LINK : fatal error LNK1104: cannot open file 'libboost_python-vc140-mt-s-1_63.lib' [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
Done Building Project "C:\dlib-19.4\tools\python\build\dlib_.vcxproj" (default targets) -- FAILED.
Done Building Project "C:\dlib-19.4\tools\python\build\ALL_BUILD.vcxproj" (default targets) -- FAILED.
Done Building Project "C:\dlib-19.4\tools\python\build\install.vcxproj" (default targets) -- FAILED.
Build FAILED.
</pre>

This message appears again at the end.
<pre>
LINK : fatal error LNK1104: cannot open file 'libboost_python-vc140-mt-s-1_63.lib' [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
    18 Warning(s)
    1 Error(s)
error: cmake build failed!
</pre>

Remark:  Looking at the c:\boost_1_63_0\stage\lib, there is another file called libboost_python3-vc140-mt-s-1_63.lib.  
Question: Is this the file libboost_python3-vc140-mt-s-1_63 rather than libboost_python-vc140-mt-s-1_63.lib. Which make file should I modify if this is the problem.

Other question. What else do I need to do and/or modify.
Please help, most grateful,

Anthony from Sydney NSW

**Comment by AnthonyTheKoala** [comment]
[results.txt](https://github.com/davisking/dlib/files/854350/results.txt)
Apologies, I have included the file results.txt which is the full output of python setup.py install -G "Visual Studio 14 2015 Win64" > results.txt

Other apology,
Where I said in point (4)
<pre>
bjam --with-python --toolset=msvc-2015 address-model=64 variant=release
</pre>
I mistakenly transcribed the commands. It should be
<pre>
bjam --with-python --toolset=msvc address-model=64 variant=release
</pre>

Thank you and most grateful for an answer,
Anthony from Sydney NSW

**Comment by davisking** [comment]
The CMake output from dlib tells you what boost commands you need to build boost if it doesn't find boost.  In particular, try this command to build all the boost files:

```
 b2 -a --with-python address-model=64 toolset=msvc runtime-link=static
```

**Comment by AnthonyTheKoala** [comment]
In response,
(1) Opened a new cmd window with admin that runs the amd64 version of CL 
(2) in c:\boost_1_63_0
<pre>
b2 -a --with-python address-model=64 toolset=msvc runtime-link=static

found 3332 targets.

lots of C4244 warnings,on Py_ssize_t to int or to unsigned int and __int64 to int, "loss of information warnings"

...updated 84 targets...

The Boost C++ were successfully built!
</pre>


(3) Opened a new cmd window with admin that runs the amd64 version of CL
(4) in c:\dlib-19.4
<pre>
python setup.py install -G "Visual Studio 14 2015 Win64"

</pre>
It still says boost not found
<pre>Call Stack (most recent call first):
  C:/dlib-19.4/dlib/cmake_utils/add_python_module:61 (FIND_PACKAGE)
  CMakeLists.txt:6 (include)
-- Could NOT find Boost
CMake Warning at C:/Program Files/CMake/share/cmake-3.8/Modules/FindBoost.cmake:1520 (message):
  No header defined for python-py35; skipping header check
Call Stack (most recent call first):
  C:/dlib-19.4/dlib/cmake_utils/add_python_module:63 (FIND_PACKAGE)
  CMakeLists.txt:6 (include)
-- Could NOT find Boost
</pre>
Yet it still mentions using boost libs
<pre>
all Stack (most recent call first):
  C:/dlib-19.4/dlib/cmake_utils/add_python_module:66 (FIND_PACKAGE)
  CMakeLists.txt:6 (include)
-- Boost version: 1.63.0
-- Found the following Boost libraries:
--   python3
-- Found PythonLibs: C:/Python34/libs/python34.lib (found suitable version "3.4.4", minimum required is "3.4")
-- USING BOOST_LIBS: optimized;C:/boost_1_63_0/stage/lib/libboost_python3-vc140-mt-s-1_63.lib;debug;C:/boost_1_63_0/stage/lib/libboost_python3-vc140-mt-sgd-1_63.lib
-- USING PYTHON_LIBS: C:/Python34/libs/python34.lib
-- C++11 activated.
</pre>

Still get the C4244 warnings 
<pre>
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\CL.exe /c /IC:\boost_1_63_0 /IC:\Python34\include /I"C:\dlib-19.4\dlib\.." /nologo /W3 /WX- /O2 /Ob2 /D WIN32 /D _WINDOWS /D NDEBUG /D BOOST_PYTHON_STATIC_LIB /D DLIB_HAVE_SSE2 /D DLIB_HAVE_SSE3 /D DLIB_HAVE_SSE41 /D DLIB_VERSION=19.4.0 /D "CMAKE_INTDIR=\"Release\"" /D dlib__EXPORTS /D _WINDLL /D _MBCS /Gm- /EHsc /MT /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"dlib_.dir\Release\\" /Fd"dlib_.dir\Release\vc140.pdb" /Gd /TP /errorReport:queue   /bigobj "C:\dlib-19.4\tools\python\src\dlib.cpp" "C:\dlib-19.4\tools\python\src\matrix.cpp" "C:\dlib-19.4\tools\python\src\vector.cpp" "C:\dlib-19.4\tools\python\src\svm_c_trainer.cpp" "C:\dlib-19.4\tools\python\src\svm_rank_trainer.cpp" "C:\dlib-19.4\tools\python\src\decision_functions.cpp" "C:\dlib-19.4\tools\python\src\other.cpp" "C:\dlib-19.4\tools\python\src\basic.cpp" "C:\dlib-19.4\tools\python\src\cca.cpp" "C:\dlib-19.4\tools\python\src\sequence_segmenter.cpp" "C:\dlib-19.4\tools\python\src\svm_struct.cpp" "C:\dlib-19.4\tools\python\src\image.cpp" "C:\dlib-19.4\tools\python\src\rectangles.cpp" "C:\dlib-19.4\tools\python\src\object_detection.cpp" "C:\dlib-19.4\tools\python\src\shape_predictor.cpp" "C:\dlib-19.4\tools\python\src\correlation_tracker.cpp" "C:\dlib-19.4\tools\python\src\face_recognition.cpp" "C:\dlib-19.4\tools\python\src\gui.cpp"
  dlib.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  matrix.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  vector.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  svm_c_trainer.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  svm_rank_trainer.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  decision_functions.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  other.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  basic.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  cca.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  sequence_segmenter.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  svm_struct.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  image.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  rectangles.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  object_detection.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  shape_predictor.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  correlation_tracker.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  face_recognition.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  gui.cpp
C:\boost_1_63_0\boost/python/detail/caller.hpp(55): warning C4244: 'return': conversion from 'Py_ssize_t' to 'unsigned int', possible loss of data [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
  Generating Code...
</pre>



It seems to get stuck for a few minutes after the gui.obj compilation and we still get the LNK1104 error
<pre>
 dlib_.dir\Release\gui.obj
LINK : fatal error LNK1104: cannot open file 'libboost_python-vc140-mt-s-1_63.lib' [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
Done Building Project "C:\dlib-19.4\tools\python\build\dlib_.vcxproj" (default targets) -- FAILED.
Done Building Project "C:\dlib-19.4\tools\python\build\ALL_BUILD.vcxproj" (default targets) -- FAILED.
Done Building Project "C:\dlib-19.4\tools\python\build\install.vcxproj" (default targets) -- FAILED.
Build FAILED.
</pre>
AND here
<pre>

LINK : fatal error LNK1104: cannot open file 'libboost_python-vc140-mt-s-1_63.lib' [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
Done Building Project "C:\dlib-19.4\tools\python\build\dlib_.vcxproj" (default targets) -- FAILED.
Done Building Project "C:\dlib-19.4\tools\python\build\ALL_BUILD.vcxproj" (default targets) -- FAILED.
Done Building Project "C:\dlib-19.4\tools\python\build\install.vcxproj" (default targets) -- FAILED.
Build FAILED.
"C:\dlib-19.4\tools\python\build\install.vcxproj" (default target) (1) ->
"C:\dlib-19.4\tools\python\build\ALL_BUILD.vcxproj" (default target) (3) ->
"C:\dlib-19.4\tools\python\build\dlib_.vcxproj" (default target) (5) ->
</pre>


Still don't know why I still get the LNK1104 error. Is there anything from my first post that I did wrong?

Attached is results.txt [results.txt](https://github.com/davisking/dlib/files/856698/results.txt)

Most grateful for an answer

Thanks
Anthony of Sydney NSW

**Comment by davisking** [comment]
No idea.  Maybe your BOOST_ROOT and BOOST_LIBRARYDIR aren't right.  You know you have to log out of windows for that to take effect right?

**Comment by AnthonyTheKoala** [comment]
Restarted the machine. Still getting the LNK1004 error.
I checked the BOOST_ROOT. It is c:\boost_1_63_0 and that appears in the the output results.txt (see previous installations). The output mentions references too BOOST_ROOT being c:\boost_1_63_0
BOOST_LIBRARYDIR: c:\boost_1_63_0\stage\lib. Examined the contents, contains 90 files, of the *.dll and *.lib type.
BOOST_INCLUDE: c:\boost_1_63_0\boost - folder contains *.h files and c++ header files. Also are subfolders including a python subfolder. It too has *.h files.

Questions:
Should BOOST_INCLUDE be set to c:\boost_1_63_0\boost\python?
Or should it be set to C:\Boost\include\boost-1_63\boost\python. Noting that the c:\Boost folder was created as a result of 'installing' Boost.
There appears to be duplications of these folders.
Documentation on boost is confusing.

Please most grateful for your assistance.

Thank you,
Anthony from Sydney NSW

**Comment by AnthonyTheKoala** [comment]
Dear Davis, 
The files do exist. In fact they exist at:
<pre>
libboost_python-vc140-mt-s-1_63.lib exists in c:\boost_1_63_0\stage\lib 
libboost_python-vc140-mt-s-1_63.lib exists also in c:\boost_1_63)9\stage 
</pre>
There is also a file 
<pre>
libboost_python-vc140-mt-s-1_63.lib.rsp
</pre>
Located at 
<pre>
C:\boost_1_63_0\bin.v2\libs\python\build\msvc-14.0\release\address-model-64\link-static\runtime-link-static\threading-multi
</pre>
The files exist. 
The issue is that the compiler outputs a LNK1104 error saying it could not open the file
<pre>
LINK : fatal error LNK1104: cannot open file 'libboost_python-vc140-mt-s-1_63.lib' [C:\dlib-19.4\tools\python\build\dlib_.vcxproj]
</pre>
At the point before the occurrence of the error, the compiler seems to 'hang up' or compiling for a few minutes before it issues the LNK1104 error.

What I don't understand is that it says in these few lines, that it could not find Boost, but then found it.
Look at the first line, then look at USING BOOST_LIBS. The program did find boost.
<pre>
-- Could NOT find Boost
CMake Warning at C:/Program Files/CMake/share/cmake-3.8/Modules/FindBoost.cmake:1520 (message):
  No header defined for python3; skipping header check
Call Stack (most recent call first):
  C:/dlib-19.4/dlib/cmake_utils/add_python_module:66 (FIND_PACKAGE)
  CMakeLists.txt:6 (include)
-- Boost version: 1.63.0
-- Found the following Boost libraries:
--   python3
-- Found PythonLibs: C:/Python34/libs/python34.lib (found suitable version "3.4.4", minimum required is "3.4")
-- USING BOOST_LIBS: optimized;C:/boost_1_63_0/stage/lib/libboost_python3-vc140-mt-s-1_63.lib;debug;C:/boost_1_63_0/stage/lib/libboost_python3-vc140-mt-sgd-1_63.lib
</pre>

If you look at the last line above, it says it is using libboost-python3-vc140-mt-sgd-1_63.lib
Based on the reference http://www.boost.org/doc/libs/1_63_0/more/getting_started/windows.html#library-naming  it says multi-threaded, static, debug, debug-python. 
But I only wanted a static multithreaded release. as in the file "'libboost_python-vc140-mt-s-1_63.lib'" that the compiler said it could not open though it existed in two locations as mentioned at the beginning of the response.

So I will do two things.
(1) remove the  BOOST_INCLUDE variable and path
(2) modify BOOST_LIBRARYDIR to be "c:\boost_1_63_0\stage\lib";"C:\boost_1_63_0\bin.v2\libs\python\build\msvc-14.0\release\address-model-64\link-static\runtime-link-static\threading-multi";
(3) restart the machine
(4) recompile dlib

Thank you and most grateful for anyone's assistance.

Anthony of Sydney NSW

**Comment by AnthonyTheKoala** [comment]
Unfortunately including the extra libraries did not help.
Still cannot open the libboost-python3-vc140-mt-sgd-1_63.lib even though it exists.
To be honest, if the cause of the problem is creating *.vcxprojl files I don't need them. If the problem is creating *.vcxprojl files, which are related to the Visual Studio IDE, I am not interested as I don't have the Visual Studio IDE. I only have the 2010 version in which its linker, compilers and other associated things are not in the "path".

Most grateful for an answer.
Anthony of Sydney NSW

**Comment by davisking** [comment]
Don't worry about visual studio seeming to hang.  It just does that all the time.  It's horribly slow even when working normally, compared to fast compilers like gcc or clang.

It says it can't find boost before it finds it because there are multiple statements in cmake that look for boost.  The first few don't find it, then it finds it.  There are multiple checks because there is no uniform name for the boost libraries so I have to make it check multiple times to see what's installed. That's fine.

Anyway, it should be working.  Maybe try an older version of boost.

**Comment by davisking** [comment]
For instance, I compile with boost_1_61_0 on windows all the time without any trouble.

**Comment by AnthonyTheKoala** [comment]
SUCCESS!
I managed to get dlib installed with the boost_1_63_0. First, compiling with the zipped version of boost_1_61_0 produced the same results as the zipped version of boost_1_63_0.

Here are the steps to replicate the success.

FIRST:
Question: How did I successfully install Boost_1_63_0 when I had LNK4004 errors with the zipped versions of Boost_1_63_0 and Boost_1_61_0?

Answer: I used the binary version of Boost_1_63_0 and I chose to install it at c:\boost_1_63_0 not the default c:\local\boost_1_63_0.
What is the name of the file and the location?
<pre>
boost_1_63_0-msvc-14.0-64.exe from https://sourceforge.net/projects/boost/files/boost-binaries/1.63.0/
</pre>
Was it a straightforward install? No.
Could you elaborate: Simply clicking did install a clean version of boost_1_63_0. However if you tried to install dlib-19.4 you would get a similar error to installing dlib with the zipped version of boost_1_63_0 and boost_1_61_0. 
What was the result without doing any more? It asked said that it could not open python27.lib. Strange isn't it!

STEP 2
So what did you do? Inspecting the folders in c:\boost_1_63_0,  there was an extra folder that was that was not installed when I tried to install the zipped version of boost* where * = 1_63_0 and 1_61_0
What was that folder?
<pre>
C:\boost_1_63_0\lib64-msvc-14.0
</pre>
What did it have?
It had the file
<pre>
libboost_python-vc140-mt-s-1_63
</pre>
which for some inexplicable reason could not open in installations of dlib described above even though the file existed.
What did you do? I included C:\boost_1_63_0\lib64-msvc-14.0 in the BOOST_LIBRARYDIR.
<pre>
BOOST_LIBRARYDIR
C:\boost_1_63_0\lib64-msvc-14.0;c:\boost_1_63_0\stage\lib;c:\boost_1_63_0\bin.v2\libs\python\build\msvc-14.0\release\address-model-64\link-static\runtime-link-static\threading-multi;c:\boost_1_63_0\bin.v2\libs\python\build\msvc-14.0\release\address-model-64\link-static\threading-multi
</pre>

STEP3
What else did you do? I did not use the existing command window. I opened a new command window or new C++ tools amd64 window. By doing this, you did not have to restart the Windows because when you said in the console SET BOOST_LIBRARYDIR, you could see the new environment variable.
Yeah, but what else did you do? I installed Boost.Python.
I went to c:\boost_1_63_0\ and did the following
<pre>
bootstrap
b2 -a --with-python address-model=64 toolset=msvc runtime-link=static
</pre>

STEP4
Then I went to c:\dlib-19.4 and did the following
<pre>
python setup.py install -G "Visual Studio 14 2014 Win64"
</pre>

RESULT? The result is that the dlib installation went smoothly. 
There were the usual C4244 warnings, and interestingly, C4005 warnings in regards to DLIB_PNG_SUPPORT and DLIB_PNG_and "macro redefinition".

NOW: I go to IDLE, and type the following
<pre>
Python 3.4.4 (v3.4.4:737efcadf5a6, Dec 20 2015, 20:20:57) [MSC v.1600 64 bit (AMD64)] on win32
Type "copyright", "credits" or "license()" for more information.
>>> import dlib as boo
>>> boo.__version__
'19.4.0'
>>> 
</pre>

My hope is that nothing crashes when I experiment. I may in future install CUDA which is big and reinstall dlib.

My question is why did the binary version have the extra folder C:\boost_1_63_0\lib64-msvc-14.0 ?

Thanks
Anthony of Sydney Australia

**Comment by davisking** [comment]
Huh, weird.  It definitely works on windows 7 with MSVC2015 if you download the boost source and run the commands I said.  That's what I do.  But who knows.  The binary folder name shouldn't matter.

I also use cuda on windows sometimes and cmake should just pick it up without you needing do anything if cuda is installed.  Although maybe you will find out otherwise.  Anyway, looks like it's working.
